common/persistence/cassandra/queue_v2_store.go

```diff
@@ -0,0 +1,242 @@
+	//  |                  |                                    |
+	//  |<--2. Return ID X--------------------------------------|
+	//  |                  |                                    |
+	//  |                  |---3. SELECT MAX(ID)--------------->|
+	//  |                  |                                    |
+	//  |                  |<--4. Return ID X-------------------|
+	//  |                  |                                    |
+	//  |---5. INSERT (X+1, msgA) with CAS--------------------->|
+	//  |                  |                                    |
+	//  |<--6. Acknowledge--------------------------------------|
+	//  |                  |                                    |
+	//  |                  |---7. INSERT (X+1, msgB) with CAS-->|
+	//  |                  |                                    |
+	//  |                  |<--8. Conflict/Error----------------|
+	//  |                  |                                    |
+	ErrQueueMessageConflict            = errors.New("queue message with id already exists, likely due to concurrent writes")
+	ErrInvalidQueueMessageEncodingType = errors.New("invalid encoding type for queue message")
+)
+
+func NewQueueV2Store(session gocql.Session) persistence.QueueV2 {
+	return &queueV2Store{
+		session: session,
+	}
+}
+
+func (q *queueV2Store) EnqueueMessage(
+	ctx context.Context,
+	request *persistence.InternalEnqueueMessageRequest,
+) (*persistence.InternalEnqueueMessageResponse, error) {
+	messageID, err := q.getNextMessageID(ctx, request.QueueType, request.QueueName)
+	if err != nil {
+		return nil, err
+	}
+
+	err = q.tryInsert(ctx, request.QueueType, request.QueueName, request.Blob, messageID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &persistence.InternalEnqueueMessageResponse{
+		Metadata: persistence.MessageMetadata{ID: messageID},
+	}, nil
+}
+
+func (q *queueV2Store) ReadMessages(
+	ctx context.Context,
+	request *persistence.InternalReadMessagesRequest,
+) (*persistence.InternalReadMessagesResponse, error) {
+	if request.PageSize <= 0 {
+		return nil, persistence.ErrNonPositiveReadQueueMessagesPageSize
+	}
+	minMessageID, err := q.getMinMessageID(request)
+	if err != nil {
+		return nil, err
+	}
+
+	iter := q.session.Query(
+		TemplateGetMessagesQuery,
+		request.QueueType,
+		request.QueueName,
+		0,
+		minMessageID,
+		request.PageSize,
+	).WithContext(ctx).Iter()
+
+	var (
+		messages        []persistence.QueueV2Message
+		messageID       int64
+		messagePayload  []byte
+		messageEncoding string
+	)
+
+	for iter.Scan(&messageID, &messagePayload, &messageEncoding) {
+		encoding, ok := enums.EncodingType_value[messageEncoding]
+		if !ok {
+			return nil, fmt.Errorf("%w: %v", ErrInvalidQueueMessageEncodingType, messageEncoding)
+		}
+
+		encodingType := enums.EncodingType(encoding)
+
+		// We need to clone the payload byte slice because its underlying array may be reused for another row.
+		payloadClone := slices.Clone(messagePayload)
```

Comment:
create a new byte array for messagePayload on each iteration, then we can avoid the clone?
or check how existing queue impl does this, I don't think we are doing clone there.

Reply:
Done